File: algorithms/searching/binary_search/golang/binary.go
package main

import "fmt"

func main() {
	numbers := []int{2, 3, 4, 10, 23, 140}
	target := 10
	result := iterativeBinarySearch(numbers, target)

	if result == -1 {
		fmt.Println("Couldn't find", target, ":(")
	} else {
		fmt.Println("Found", target, "at index", result, "!")
	}
}

func iterativeBinarySearch(numbers []int, target int) int {
	min := 0
	max := len(numbers) - 1
	mid := len(numbers) / 2

	for min <= max {
		value := numbers[mid]
		if value == target {
			return min
		}
		if value > target {
			max = mid - 1
			mid = (min + max) / 2
			continue
		}
		min = mid + 1
		mid = (min + max) / 2
	}
	return -1
}

File: algorithms/searching/binary_search/binary.dart
void main() {
  List<int> numbers = [2, 3, 4, 10, 20];
  int target = 10;
  int min = 0;
  int max = numbers.length - 1;
  int result = binarySearch(numbers, target, min, max);
  if (result == -1){
    print("Couldn't find $target :(");
  } else{
    print("Found $target at index $result!");
  }
}

int binarySearch(List<int> numbers, int target, int min, int max) {
  if (max >= min) {
    int mid = ((max + min) / 2).floor();
    if (numbers[mid] > target){
      return binarySearch(numbers, target, min, mid);
    } else if(numbers[mid] < target){
      return binarySearch(numbers, target, mid+1, max);
    }else{
      return mid;
    }
  }
  return -1;
}
